File: aoc2023/day4b.py
import sys


def run():
    n_matches: list[int] = []
    n_cards: dict[int, int] = {}

    for line in sys.stdin:
        split_line = line.split("|")
        memo = set(
            [int(x) for x in split_line[0].split(":")[1].split(" ") if x.isnumeric()]
        )

        numbers = [int(x) for x in split_line[1].split() if x.isnumeric()]
        matches = sum([x in memo for x in numbers])
        n_matches.append(matches)

    def dp(index: int) -> int:
        if index in n_cards:
            return n_cards[index]
        result = 1
        for i in range(n_matches[index]):
            result += dp(index + i + 1)
        n_cards[index] = result
        return result

    result = 0
    for i in range(len(n_matches) - 1, -1, -1):
        result += dp(i)

    print(result)


if __name__ == "__main__":
    run()

File: src/aoc2023/day4b.rs
use std::collections::HashMap;
use std::collections::HashSet;
use std::io::BufRead;

pub fn run() {
    let matches: Vec<u32> = std::io::stdin()
        .lock()
        .lines()
        .map(|line| {
            let line = line.unwrap();
            let split_line: Vec<&str> = line.split("|").collect();

            let mut memo: HashSet<&str> = HashSet::new();
            split_line[0]
                .split(":")
                .last()
                .unwrap()
                .split_whitespace()
                .for_each(|x| {
                    memo.insert(x);
                });

            let matches: u32 = split_line[1]
                .split_whitespace()
                .map(|x| match memo.contains(x) {
                    true => 1,
                    false => 0,
                })
                .sum();

            return matches;
        })
        .collect();

    let mut memo: HashMap<usize, u32> = HashMap::new();
    fn dp(index: usize, memo: &mut HashMap<usize, u32>, matches: &Vec<u32>) -> u32 {
        if memo.contains_key(&index) {
            return *memo.get(&index).unwrap();
        } else {
            let mut result: u32 = 1;
            for i in 1..=matches[index] {
                result += dp(index + i as usize, memo, matches);
            }
            memo.insert(index, result);

            return result;
        }
    }

    let mut result: u32 = 0;

    for i in 0..matches.len() {
        result += dp(i, &mut memo, &matches);
    }

    println!("{result}");
}
